File: frontend/src/play/company.ts
import type { GatheringView, Guest, Tint } from "../api/gathering";

/** What a guest standing at the gathering is told, since the players of the game are the ones who settle it. */
const STANDING = "Take a seat to deal";

/** Every seat of the table the company settled on, which is what a page draws a place for. */
export function seatsOf(gathering: GatheringView): number[] {
  return [...Array(gathering.choice.players).keys()];
}

/** The guest holding one seat, and nothing where it stands empty. */
export function holderOf(gathering: GatheringView, seat: number): Guest | null {
  return gathering.company.find((guest) => guest.seat === seat) ?? null;
}

/** The guests at the gathering who hold no seat, which is everyone watching it. */
export function standingBy(gathering: GatheringView): Guest[] {
  return gathering.company.filter((guest) => guest.seat === null);
}

/** The seats of the table nobody holds, which are the ones the deal waits on. */
export function emptySeats(gathering: GatheringView): number[] {
  return seatsOf(gathering).filter((seat) => holderOf(gathering, seat) === null);
}

/** The guest reading the page, and nothing where the company reads nobody by that name. */
export function me(gathering: GatheringView): Guest | null {
  return gathering.company.find((guest) => guest.name === gathering.mine) ?? null;
}

/** The seat the guest reading the page holds, and nothing while they are standing. */
export function mySeat(gathering: GatheringView): number | null {
  return me(gathering)?.seat ?? null;
}

/** The guests holding a seat, who are the players the deal is dealt to and the ones a commitment is asked of. */
export function seatedGuests(gathering: GatheringView): Guest[] {
  return gathering.company.filter((guest) => guest.seat !== null);
}

/** Whether the guest reading the page gathered the table, whose say governs it while it is settled host by host. */
export function iAmHost(gathering: GatheringView): boolean {
  return me(gathering)?.host ?? false;
}

/** Whether the guest reading the page has committed to the settings as they stand. */
export function iAmReady(gathering: GatheringView): boolean {
  return me(gathering)?.ready ?? false;
}

/**
 * Whether the guest reading the page may call the deal.
 *
 * A democratic table leaves the deal to every seat, as it leaves the settling; a host-governed one keeps it to
 * the host alone, so a seated guest reads it as the wait it is until the host calls it.
 */
export function mayDeal(gathering: GatheringView): boolean {
  return gathering.democratic || iAmHost(gathering);
}

/** Whether every seat is taken and every seated guest has committed, which is the whole of what the deal waits on. */
export function everyoneReady(gathering: GatheringView): boolean {
  return emptySeats(gathering).length === 0 && seatedGuests(gathering).every((guest) => guest.ready);
}

/** Whether this seat is the one the guest reading the page holds. */
export function mine(gathering: GatheringView, seat: number): boolean {
  return mySeat(gathering) === seat;
}

/** The tint the guest reading the page plays under, and nothing where the company reads nobody by that name. */
export function myTint(gathering: GatheringView): Tint | null {
  return gathering.company.find((guest) => guest.name === gathering.mine)?.tint ?? null;
}

/** The guest holding one tint, and nothing where it stands free for whoever takes it. */
export function tintHeldBy(gathering: GatheringView, tint: Tint): Guest | null {
  return gathering.company.find((guest) => guest.tint === tint) ?? null;
}

/**
 * Whether the guest reading the page holds a say over what the table plays and when it is dealt.
 *
 * This mirrors `cardserver.gathering.SeatedSay`: standing at a gathering is watching it and taking a seat is
 * joining the game, so the players of the game settle what is played. The page reads it to draw its controls
 * as the server would answer them, and the server answers for itself either way.
 */
export function hasSay(gathering: GatheringView): boolean {
  return mySeat(gathering) !== null;
}

/**
 * What stands in the way of the deal, and nothing where the table may be dealt as it stands.
 *
 * A gathering is dealt once every seat is taken and every seated guest has committed to the settings, and by a
 * guest holding a seat at it. Each is what the server refuses on, so what is read here is the sentence a person
 * is shown before they meet the refusal: a seat still to be taken first, and a commitment still to be given once
 * the table is full.
 */
export function holdingUpTheDeal(gathering: GatheringView): string | null {
  if (!hasSay(gathering)) {
    return STANDING;
  }

  const empty = emptySeats(gathering);
  if (empty.length > 0) {
    return empty.length === 1 ? "One seat still to be taken" : `${empty.length} seats still to be taken`;
  }

  const waiting = seatedGuests(gathering).filter((guest) => !guest.ready).length;
  if (waiting > 0) {
    return waiting === 1 ? "One player still to ready" : `${waiting} players still to ready`;
  }

  return null;
}

/**
 * The one press the room carries a seated guest through, from committing to the settings to calling the deal.
 *
 * `stand` is for the guest holding no seat, who neither commits nor deals. `ready` is for a seated guest while
 * anything holds the deal up: they commit to the settings or take that commitment back, and `pending` reads what
 * the table still waits on. `deal` stands once nothing does and the guest may call it. `wait` stands then for a
 * seated guest a host-governed table leaves the deal to the host: they are ready and the table is, and the press
 * is the host's to make.
 */
export type Commit =
  | { readonly act: "stand" }
  | { readonly act: "ready"; readonly ready: boolean; readonly pending: string }
  | { readonly act: "deal" }
  | { readonly act: "wait" };

/** How the room's one press reads for the guest reading it, out of where the gathering and their seat stand. */
export function commitOf(gathering: GatheringView): Commit {
  if (!hasSay(gathering)) {
    return { act: "stand" };
  }

  const pending = holdingUpTheDeal(gathering);
  if (pending !== null) {
    return { act: "ready", ready: iAmReady(gathering), pending };
  }

  return mayDeal(gathering) ? { act: "deal" } : { act: "wait" };
}

File: tests/server/company.py
from collections.abc import Mapping
from dataclasses import dataclass
from random import Random
from typing import Final

from fastapi import FastAPI

from cardserver.app import create_app
from cardserver.gathering import Gathering, Gatherings, GovernedSay, Turnstile
from cardserver.naming import Named, Seated
from cardserver.protocols import TableId
from cardserver.registry import TableRegistry
from cardserver.remembering import RoomRecord, TableRecord
from cardserver.schemas import Choice, Offering
from cardwork.games.capacity import Capacity
from cardwork.moves.actions import Play
from cardwork.moves.move import Move
from cardwork.rounds.conclusion import Conclusion

from ..games.demo import DECK, SealedRoundGame
from .keeping import Keeping, a_journal, applied_of
from .layout import SEALED_SCENE, TITLE

GAME: Final[str] = "sealed"
OTHER_GAME: Final[str] = "unsealed"
OTHER_TITLE: Final[str] = "Unsealed round"
CODE: Final[str] = "KQAJ72"
WRONG_CODE: Final[str] = "234567"
COMPANY_LEAST: Final[int] = 2
COMPANY_MOST_HERE: Final[int] = 4
ONE_DECK: Final[int] = 1
TWO_DECKS: Final[int] = 2
ROUNDS: Final[int] = 1
NO_GRACE: Final[float] = 0.0
DEALT_FROM: Final[int] = 20260806
TURNSTILE_WINDOW: Final[float] = 60.0
WRONG_CODES_ALLOWED: Final[int] = 10
SWEEP_SECONDS: Final[float] = 60.0
STREAM_PATIENCE: Final[float] = 0.5
PRESENCE_STANDS: Final[float] = 60.0

OFFERINGS: Final[tuple[Offering, ...]] = (
    Offering(
        game=GAME,
        title=TITLE,
        seats=Capacity(least=COMPANY_LEAST, most=COMPANY_MOST_HERE),
        decks=(ONE_DECK,),
    ),
    Offering(
        game=OTHER_GAME,
        title=OTHER_TITLE,
        seats=Capacity(least=COMPANY_LEAST, most=COMPANY_MOST_HERE),
        decks=(ONE_DECK, TWO_DECKS),
    ),
)


COMPANY: Final[tuple[str, ...]] = ("Ada", "Grace", "Alan")
FIRST_CARD: Final[frozenset[int]] = frozenset({0})


def a_sealing(seat: int) -> Move:
    """One seat sealing the first card of its hand in its own tray, which is the move this game is played by."""
    return Move(player=seat, action=Play(group="sealed", indices=FIRST_CARD))


def a_sealed_round(players: int) -> Choice:
    """A choice of the demo game at a table of that many seats, run to a single round."""
    return Choice(
        game=GAME,
        players=players,
        decks=ONE_DECK,
        conclusion=Conclusion(rounds=ROUNDS),
    )


class Ticking:
    """A clock a test moves by hand, which is what lets a window pass without a test waiting for it."""

    def __init__(self) -> None:
        self._now = 0.0

    def __call__(self) -> float:
        return self._now

    def on(self, seconds: float) -> None:
        """Move the clock forward, which is how long a test says has gone by."""
        self._now += seconds


class Deals:
    """An opening that deals the demo game, which is what a gathering at these tests becomes.

    What it was handed is kept as well as opened, so a test reads the seating a deal settled straight off it,
    and reads it through a layout where the point is that the names and tints reach one.
    """

    def __init__(self, registry: TableRegistry) -> None:
        self._registry = registry
        self.dealt: list[tuple[TableId, Choice, Mapping[int, Seated]]] = []
        self.resumed: list[TableId] = []

    def open(
        self,
        table: TableId,
        choice: Choice,
        seated: Mapping[int, Seated],
    ) -> None:
        """Deal the demo game at the table the choice settled, with its seats named and tinted."""
        self.dealt.append((table, choice, seated))
        self._registry.open(
            table,
            self._match(choice),
            Named(SEALED_SCENE, seated),
        )

    def resume(self, room: RoomRecord, record: TableRecord) -> None:
        """Take one table back up from the record kept of it, which is what a run does over what it wrote down.

        The rules are handed the record and stand where its last commit left them, and the keys the lines
        carry go to the session, so a retry crossing the restart is answered with the sequence it reached.
        Whatever the window a run was cut off inside would have committed is settled here and written down as
        the table opens, since a window belongs to the process that opened it.
        """
        game = self._match(room.choice)
        game.resume(a_journal(record))
        settled = game.settle()
        self.resumed.append(room.table)
        self._registry.reopen(
            room.table,
            game,
            Named(SEALED_SCENE, room.seated),
            applied=applied_of(record),
            settled=settled,
        )

    def _match(self, choice: Choice) -> SealedRoundGame:
        """A match of the demo game at the table the choice settled, dealt from the deck these tests play with."""
        return SealedRoundGame(
            players=choice.players,
            deck=DECK,
            rng=Random(DEALT_FROM),
        )


@dataclass(frozen=True)
class Rig:
    """One run under test: the tables it serves, the rooms it gathers, the store it writes to and its clock.

    A run gathering a fresh table and a run reading one back from a record hold the same things, so this is
    what both of them start from and a restart under test is a second rig over the store the first wrote to.

    The application carries the routes a table is played at as well, since a gathering is dealt into a table
    the same application answers for: reading a plaque after the deal is what shows a name reaching the felt.
    """

    registry: TableRegistry
    gatherings: Gatherings
    deals: Deals
    keeping: Keeping
    ticking: Ticking
    app: FastAPI


@dataclass(frozen=True)
class Gathered:
    """One table gathering under test, read together with the run that holds it."""

    registry: TableRegistry
    gatherings: Gatherings
    gathering: Gathering
    deals: Deals
    keeping: Keeping
    ticking: Ticking
    app: FastAPI


def a_rig(keeping: Keeping) -> Rig:
    """A run over one store, holding an empty lobby until a table is gathered at it or read back into it."""
    ticking = Ticking()
    registry = TableRegistry(NO_GRACE, keeping=keeping, clock=ticking)
    deals = Deals(registry)
    gatherings = Gatherings(
        deals,
        offerings=OFFERINGS,
        say=GovernedSay(),
        turnstile=Turnstile.watching(
            ticking,
            window=TURNSTILE_WINDOW,
            wrong_codes_allowed=WRONG_CODES_ALLOWED,
        ),
        keeping=keeping,
        clock=ticking,
        presence_stands=PRESENCE_STANDS,
    )
    return Rig(
        registry=registry,
        gatherings=gatherings,
        deals=deals,
        keeping=keeping,
        ticking=ticking,
        app=create_app(
            registry,
            gatherings,
            gatherings,
            sweep_seconds=SWEEP_SECONDS,
            stream_patience=STREAM_PATIENCE,
        ),
    )


def at(rig: Rig, gathering: Gathering) -> Gathered:
    """One room of a run's lobby, read together with the run holding it."""
    return Gathered(
        registry=rig.registry,
        gatherings=rig.gatherings,
        gathering=gathering,
        deals=rig.deals,
        keeping=rig.keeping,
        ticking=rig.ticking,
        app=rig.app,
    )


def a_seated_company(gathered_at: Gathered) -> dict[str, str]:
    """The whole company arrived, seated in order and committed, and the token each of them speaks through."""
    gathering = gathered_at.gathering
    tokens = {name: gathering.admit(name) for name in COMPANY}
    for seat, name in enumerate(COMPANY):
        gathering.claim(name, seat, gathering.revision)

    for name in COMPANY:
        gathering.ready(name, True, gathering.revision)

    return tokens


def a_dealt_table(gathered_at: Gathered) -> dict[str, str]:
    """The table under test dealt into service, and the token each of the company plays through."""
    tokens = a_seated_company(gathered_at)
    gathered_at.gathering.deal(gathered_at.gathering.revision)
    return tokens


def gathered(table: TableId, players: int) -> Gathered:
    """One table gathering on its code, at a choice of the demo game seating that many."""
    rig = a_rig(Keeping())
    return at(
        rig,
        rig.gatherings.open(
            table,
            CODE,
            a_sealed_round(players),
            democratic=True,
        ),
    )


def restarted(gathered_at: Gathered) -> Rig:
    """A fresh run over everything an earlier one wrote down, holding the rooms and tables its records hold.

    This is the whole of a restart as the adapter meets it: nothing the first run held in memory carries over,
    only what its store was told, and the second run gathers its lobby from that. Rooms come back first and
    tables after, and a journal written down is the mark a deal left, whatever the room says of itself.
    """
    rig = a_rig(gathered_at.keeping)
    for kept in gathered_at.keeping.kept():
        rig.gatherings.restore(kept.room, dealt=kept.table is not None or kept.room.dealt)

    for kept in gathered_at.keeping.kept():
        if kept.table is not None:
            rig.deals.resume(kept.room, kept.table)

    return rig
